driver/generic/xpt2046.py

```diff
@@ -0,0 +1,123 @@
+        '''
+        self.buf = bytearray(3)
+        self.spi = spi
+        self.cs  = (machine.Pin(cs,machine.Pin.OUT) if isinstance(cs,int) else cs)
+        self.cs.value(1)
+        if bits not in (8,12): raise ValueError('Xpt2046.bits: must be 8 or 12 (not %s)'%str(bits))
+        self.conv=(Xpt2046.CONV_8_BIT if bits==8 else Xpt2046.CONV_12_BIT)
+        self.xy_range,self.dim,self.rot=ranges,(width,height),(rot%4)
+        self.xy_scale=[self.dim[ax]*1./(self.xy_range[ax][1]-self.xy_range[ax][0]) for ax in (0,1)]
+        self.xy_origin=[self.xy_range[ax][0] for ax in (0,1)]
+    def _raw2px(self,rxy):
+        'Convert raw coordinates to pixel coordinates'
+        x,y=[int(self.xy_scale[ax]*(rxy[ax]-self.xy_origin[ax])) for ax in (0,1)]
+        if   self.rot==0: return self.dim[0]-x,y
+        elif self.rot==1: return self.dim[1]-y,x
+        elif self.rot==2: return x,self.dim[1]-y
+        else:             return y,self.dim[0]-x
+    def _raw_pos(self):
+        'Read raw position; return value if within valid ranges (`__init__(ranges=...)`) or `None` if outside.'
+        ret=[0,0]
+        for ax,chan in [(0,Xpt2046.CHAN_X),(1,Xpt2046.CHAN_Y)]:
+            r=self._chanRead(chan)
+            if not self.xy_range[ax][0]<=r<=self.xy_range[ax][1]: return None
+            ret[ax]=r
+        return ret
+
+    def pos(self,N=10,attempts=20):
+        ''''
+        Get N position readings (limited by 20 attempts) and return mean position of valid readings.
+        If attempts are exhausted, return None.
+        '''
+        N,attempts=10,20
+        xx,yy,done=0,0,0
+        for _ in range(attempts):
+            if (r:=self._raw_pos()) is None: continue
+            xx+=r[0]; yy+=r[1]; done+=1
+            if done==N: break
+        else: return None
+        mx,my=xx*1./N,yy*1./N
+        return self._raw2px((mx,my))
+
+
+class Xpt2046(Xpt2046_hw):
+    lcd=None
+    spi=None
+    def indev_drv_read_cb(self, indev_drv, data):
+        # wait for DMA transfer (if any) before switchint SPI to 1 MHz
+        if self.lcd: self.lcd._rp2_wait_dma()
+        # print('.',end='')
+        if self.spiRate: self.spi.init(baudrate=1_000_000)
+        pos=self.pos()
+        if pos is None: data.state=0
+        else: (data.point.x,data.point.y),data.state=pos,1
+        # print('#',end='')
+        # switch SPI back to 24 MHz for the display
+        if self.spiRate: self.spi.init(baudrate=self.spiRate)
+        return False
+
+    def __init__(self,spi,spiRate=24_000_000,lcd=None,**kw):
+        '''XPT2046 touchscreen driver for LVGL; cf. documentation of :obj:`Xpt2046_hw` for the meaning of parameters being passed.
+
+        *lcd*: optional St77xx instance; in case of port-specific async DMA transfer on RP2, it will wait for the transfer to finish before using the SPI bus.
```

Comment:
xpt2046 is not really related to st77xx and can used with other displays, so I don't think it's clean to pass st77xx lcd parameter here. An input device driver should be completely independent of the display driver.  

I understand you want to wait until DMA transfer is finished, and this doesn't have to be specific to st77xx. Instead of passing the lcd you could pass an optional callback parameter that waits for SPI to be ready after DMA finished (something like `wait_for_spi` that returns when SPI is ready). Such callback could potentially be useful with other displays in the future.

Reply:
It might seem a bit weird at first, but on the board I used, the TSC and the LCD were connected to the same SPI. The LCD works at full speed (>10MHz), but the TSC doesn't work faster than (about 1MHz), so I had to reconfigure it every time I wanted to read the TSC. It is also necessary to wait for the DMA to finish to reconfigure the SPI for same reason.

If I had designed the hardware, I would have preferred a dedicated SPI for the LCD and a more "independent" sw, but many boards share it since the RP2040 doesn't have too many pins.

Reply:
As a side note, there does not seem to be API to query the *current* SPI rate (to read it, set to 1MHz for reading from xpt2046, then set it back), that's why `Xpt2046.spiRate` is needed to store the desired non-touch SPI rate.

Reply:
I am passing `spiPrereadCb` now, good idea.